metricflow/model/validations/validator_helpers.py

```diff
@@ -62,6 +65,59 @@ def context_str(self) -> str:
 
         return context_string
 
+    @staticmethod
+    def from_dict(context_dict: ValidationContextJSON) -> ValidationContext:
+        """Constructs a ValidationContext object from a JSON format dict"""
+
+        file_name = context_dict.get("file_name", None)
+        line_number = context_dict.get("line_number", None)
+
+        keys = context_dict.keys()
+        if "data_source_name" in keys:
+            data_source_name = context_dict["data_source_name"]
+
+            if "measure_name" in keys:
+                return MeasureContext(
+                    file_name=file_name,
+                    line_number=line_number,
+                    data_source_name=data_source_name,
+                    measure_name=context_dict["measure_name"],
+                )
+            elif "dimension_name" in keys:
+                return DimensionContext(
+                    file_name=file_name,
+                    line_number=line_number,
+                    data_source_name=data_source_name,
+                    dimension_name=context_dict["dimension_name"],
+                )
+            elif "identifier_name" in keys:
+                return IdentifierContext(
+                    file_name=file_name,
+                    line_number=line_number,
+                    data_source_name=data_source_name,
+                    identifier_name=context_dict["identifier_name"],
+                )
+            else:
+                return DataSourceContext(
+                    file_name=file_name,
+                    line_number=line_number,
+                    data_source_name=data_source_name,
+                )
+        elif "materialization_name" in keys:
+            return MaterializationContext(
+                file_name=file_name,
+                line_number=line_number,
+                materialization_name=context_dict["materialization_name"],
+            )
+        elif "metric_name" in keys:
+            return MetricContext(
+                file_name=file_name,
+                line_number=line_number,
+                metric_name=context_dict["metric_name"],
+            )
+        else:
+            return ValidationContext(file_name=file_name, line_number=line_number)
```

Comment:
Why is all of this necessary? Pydantic provides parse_obj for this purpose. If the model is serialized using the Pydantic `.json()` family of methods, it should be deserializable using `.parse_raw()` or `.parse_obj()`

The primary issue here seems to be the tagged union model, where you don't actually know the subclass type without inspecting the JSON value. There are two options that I think would be a bit more palatable:

1. Define a Pydantic model for the encapsulating object that includes a properly typed entry for the issue, and call parse_obj from that. It's entirely possible that my earlier suggestion to move away from a Union type alias was, for this reason, absolutely the wrong call, because you may find you need a Union type for Pydantic to work properly here.
2. Streamline this logic to call `parse_obj` once you've done your subclass detection, and then define this as a custom "validator" and override `__get_validators__` in the ValidationIssue class. This will allow callers to use the standard Pydantic APIs and still get correct results.

See https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__ for details. Also I have an example in #129

Reply:
Oooh, this looks potentially useful:

https://pydantic-docs.helpmanual.io/usage/types/#nested-discriminated-unions

Reply:
It might be the Seattle heat + time of day, but I'm struggling to put together. I think I'm most of the way there, but I'm not sure how `__get_validators__` actually gets called in the end. Once I have `__get_validators__` that returns the method which calls the correct `<class>.parse_obj`, how do I then go from a json string to the parsed object class? Calling just `ValidationIssue(json_string)` doesn't do it, and `ValidationIssue.from_raw(json_string)` seems to skip `__get_validators__` which makes sense. But then to what do I hand the json_string to be parsed? 😖